exonum/src/blockchain/block.rs

```diff
@@ -17,8 +17,55 @@ use crate::{
     crypto::Hash,
     helpers::{Height, ValidatorId},
     messages::{Precommit, Verified},
-    proto,
+    proto::{self, BinaryMap},
 };
+use exonum_merkledb::BinaryValue;
+use std::borrow::Cow;
+
+/// Trait that represents key in block header entry map. Provide
+/// mapping between `NAME` of the entry and its value.
+///
+/// # Usage
+///
+/// ```no_run
+/// use exonum::blockchain::Block;
+///
+/// struct SomeData {}
+///
+/// impl BlockHeaderKey for SomeData {
+///    const NAME: &'static str = "data";
+///    type Value = Self;
+/// }
+///
+/// let mut block = Block::default();
+///
+/// let data = SomeData {};
+/// block.insert::<SomeData>(data);
+/// ```
+pub trait BlockHeaderKey {
+    const NAME: &'static str;
+    type Value: BinaryValue;
+}
+
+/// Expandable set of entries allowed to be added to the block.
+pub type BlockHeaderEntries = BinaryMap<String, Vec<u8>>;
+
+impl BlockHeaderEntries {
+    /// New instance of `BlockHeaderEntries`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Insert new entry to the map.
+    pub fn insert<K: Into<String>, V: BinaryValue>(&mut self, key: K, value: V) {
```

Comment:
Well, since we've introduced a type dedicated for block header keys, maybe enforce its usage everywhere?
Currently it's a bit weird: `entries` is public within `Block`, and via `Block::get` we can work only with `BlockHeaderKey` types, but if we take the `entries` field from it, then the restriction is lowered to `Into<String>`.

BTW, I like the idea of separate type carved in stone, since it gently hints user that block header isn't a place for some sort of volatile data, despite of the change flexibility.